src/state/slices/opportunitiesSlice/resolvers/osmosis/utils.ts

```diff
@@ -0,0 +1,199 @@
+  return (
+    pool.hasOwnProperty('@type') &&
+    pool.hasOwnProperty('address') &&
+    pool.hasOwnProperty('id') &&
+    pool.hasOwnProperty('pool_params') &&
+    !pool.hasOwnProperty('pool_liquidity') &&
+    pool.hasOwnProperty('future_pool_governor') &&
+    pool.hasOwnProperty('total_shares') &&
+    pool.hasOwnProperty('pool_assets') &&
+    pool.hasOwnProperty('total_weight')
+  )
+}
+
+export const generateAssetIdFromOsmosisDenom = (denom: string): AssetId => {
+  if (denom.startsWith('u') && denom !== 'uosmo') {
+    return toAssetId({
+      assetNamespace: 'native' as const,
+      assetReference: denom,
+      chainId: osmosisChainId,
+    })
+  }
+
+  if (denom.startsWith('ibc')) {
+    return toAssetId({
+      assetNamespace: 'ibc' as const,
+      assetReference: denom.split('/')[1],
+      chainId: osmosisChainId,
+    })
+  }
+
+  return toAssetId({
+    assetNamespace: 'slip44' as const,
+    assetReference: ASSET_REFERENCE.Osmosis,
+    chainId: osmosisChainId,
+  })
+}
+
+export const getPools = async (): Promise<OsmosisPool[]> => {
+  try {
+    /* Fetch Osmosis pool data */
+    const { data: poolData } = await memoize(async (): Promise<AxiosResponse<OsmosisPoolList>> => {
+      return await axios.get<OsmosisPoolList>(
+        (() => {
+          const url = new URL('gamm/v1beta1/pools', getConfig().REACT_APP_OSMOSIS_LCD_BASE_URL)
+          url.searchParams.set(
+            'pagination.limit',
+            getConfig().REACT_APP_OSMOSIS_POOL_PAGINATION_LIMIT.toString(),
+          )
+          return url.toString()
+        })(),
+      )
+    })()
+
+    if (!poolData) throw new Error('Unable to fetch Osmosis liquidity pool metadata')
+
+    /* Fetch historical data for Osmosis pools */
+    const { data: historicalDataByPoolId } = await memoize(async (): Promise<any> => {
+      return await axios.get<PoolHistoricalDataList>(
+        (() => {
+          const url = new URL('pools/v2/all', getConfig().REACT_APP_OSMOSIS_IMPERATOR_BASE_URL)
+          url.searchParams.set(
+            'low_liquidity',
+            getConfig().REACT_APP_OSMOSIS_ALLOW_LOW_LIQUIDITY_POOLS.toString(),
+          )
+          return url.toString()
+        })(),
+      )
+    })()
+
+    if (!historicalDataByPoolId)
+      throw new Error('Unable to fetch historical data for Osmosis liquidity pools')
+
+    const getPoolTVL = (pool: OsmosisBasePool): string => {
+      const marketData = historicalDataByPoolId[pool.id]
+      return bnOrZero(marketData[0].liquidity).toString()
+    }
+
+    const calculatePoolAPY = (pool: OsmosisBasePool): string => {
+      const poolHistoricalData = historicalDataByPoolId[pool.id][0] // Identical pool-level historical data exists on both asset entries
+
+      /* Pool fee data is represented in API response like '0.2%'. */
+      const feeMultiplier = bnOrZero(poolHistoricalData.fees.split('%')[0]).multipliedBy(
+        bnOrZero(0.01),
```

Comment:
What's the magic `0.01` here? Also, `bn` will suffice in this instance as we are explicitly defining the input arg with a numeric literal.